src/UglyToad.PdfPig/Images/ColorSpaceDetailsByteConverter.cs

```diff
@@ -0,0 +1,99 @@
+﻿namespace UglyToad.PdfPig.Images
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Content;
+    using Core;
+    using Graphics.Colors;
+
+    /// <summary>
+    /// Utility for working with the bytes in <see cref="IPdfImage"/>s and converting according to their <see cref="ColorSpaceDetails"/>.s
+    /// </summary>
+    public static class ColorSpaceDetailsByteConverter
+    {
+        /// <summary>
+        /// Converts the output bytes (if available) of <see cref="IPdfImage.TryGetBytes"/>
+        /// to actual pixel values using the <see cref="IPdfImage.ColorSpaceDetails"/>. For most images this doesn't
+        /// change the data but for <see cref="ColorSpace.Indexed"/> it will convert the bytes which are indexes into the
+        /// real pixel data into the real pixel data.
+        /// </summary>
+        public static byte[] Convert(ColorSpaceDetails details, IReadOnlyList<byte> decoded)
+        {
+            if (decoded == null)
+            {
+                return EmptyArray<byte>.Instance;
+            }
+
+            if (details == null)
+            {
+                return decoded.ToArray();
+            }
+
+            switch (details)
+            {
+                case IndexedColorSpaceDetails indexed:
+                    return UnwrapIndexedColorSpaceBytes(indexed, decoded);
+            }
+
+            return decoded.ToArray();
+        }
+
+        private static byte[] UnwrapIndexedColorSpaceBytes(IndexedColorSpaceDetails indexed, IReadOnlyList<byte> input)
+        {
+            var multiplier = 1;
+            Func<byte, IEnumerable<byte>> transformer = null;
+            switch (indexed.BaseColorSpaceDetails.Type)
+            {
+                case ColorSpace.DeviceRGB:
+                    transformer = x =>
```

Comment:
it seems to me like this sort of code should be on an abstract method on the `ColorSpaceDetails` class :)

Reply:
Yes, I'm sort of on the fence about it, in general I'm trying to aim for keeping data and behaviour entirely separate, so data classes just encapsulate their data and enforce invariants/immutability while classes like this contain the behaviour as pure functions.

In this case though it probably makes more sense to keep the behaviour with the data, which is more traditional C# style, but I worry about ending up with a traditional C# codebase with that approach 😆